walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,231 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Collections;
+import java.util.ArrayList;
+import java.util.HashMap;
 
 public final class FindMeetingQuery {
+
+  /**
+   * This method checks if an event contains no mandatory attendees using a MeetingRequest
+   * @param request MeetingRequest object which contains list of attendees
+   * @param event The Event object's attendees are compare with the request object
+   * @return boolean if the provide event object contains no mandatory attendees
+   */
+  private boolean hasNoMandatoryAttendees(MeetingRequest request, Event event){
+    return Collections.disjoint(event.getAttendees(), request.getAttendees());
+  }
+
+  /**
+   * This method checks if an event contains only optional Attendees using a MeetingRequest
+   * @param request MeetingRequest object which contains list of optional attendees
+   * @param event The Event object's attendees are compare with the request object
+   * @return boolean if the provide event object contains only optional attendees
+   */
+  private boolean hasOnlyOptionalAttendees(MeetingRequest request, Event event){
+
+    for (String attendee : event.getAttendees()){
+      if (request.getAttendees().contains(attendee)){
+        return false;
+      }
+      if (!(request.getOptionalAttendees().contains(attendee))){
+        return false;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * This method returns the correct time ranges if optional attendees are taken into account
+   * @param optionalEvents List of events with only optional attendees
+   * @param timeRanges A list of available TimeRanges without taken into account optional events 
+   * @return A list of TimeRanges depending on the optionalEvents list and final size of timeRanges
+   */
+  private Collection<TimeRange> resolveOptionalEventConflicts(ArrayList<Event> optionalEvents, 
+        Collection<TimeRange> timeRanges){
+
+    Collection<TimeRange> backUpTimeRanges = new ArrayList<TimeRange>(timeRanges);
+    ArrayList<TimeRange> removedTimeRanges = new ArrayList<TimeRange>();
+
+    for (Event opEvent : optionalEvents){
```

Comment:
I think that you're removing an available time range from consideration if it "overlaps" with an optional event? 

Consider the following case: 
- `timeRanges = [TimeRange.WHOLE_DAY]`
- optional events contains just one event from 8:00 - 8:30. 
- request.duration is 30 minutes

Though the optional event does overlap with the single available time range, you wouldn't want to completely remove the available time range, right? 
Instead, the correct response (where all optional attendees can attend) would be [0:00 - 8:00, 8:30 - 24:00].

I think that you can update this function to handle this - let me know if you have any questions! 
Also, feel free to come back and address this in a follow-up PR.

Reply:
Hi Vivek, just to clarify, this scenario would look something like this: 
```  
// timeRanges:    |----A-----------|
// Optional: |--B--|
// Day     : |---------------------|
// Results :       |--1------------|     
```
Where ```1``` is the new ```A```

Reply:
Hi Vivek, I went over the my code and I think this is going to take a lot of time to implement. I think I would maybe like to come at a later time. Is there maybe a simple way to implement this check?

Reply:
I think if they overlap like in the picture above I would have to merge the ```TimeRanges```

Reply:
Ok, I added an Issue about these overlapping optional events checks. 

Issue: https://github.com/MauroGuerrero20/STEP_Internship_Portfolio/issues/32#issue-641519433